sdk/python/lib/pulumi/_types.py

```diff
@@ -0,0 +1,433 @@
+
+    If the class is not a subclass of dict and doesn't have an __init__()
+    method, an __init__() method is added to the class that accepts a dict
+    representing the outputs.
+    """
+
+    if is_input_type(cls) or is_output_type(cls):
+        raise AssertionError("Cannot apply @input_type and @output_type more than once.")
+
+    # Get the output properties and mark the class as an output type.
+    props = _process_class(cls, _PULUMI_OUTPUT_TYPE)
+
+    # Add an __init__() method that takes a dict (representing outputs) as an arg,
+    # if the class isn't a subclass of dict and doesn't have an __init__() method.
+    if not issubclass(cls, dict) and "__init__" not in cls.__dict__:
+        def init(self, value: dict) -> None:
+            if not isinstance(value, dict):
+                raise TypeError('Expected value to be a dict')
+            setattr(self, _VALUES, value)
+        setattr(cls, "__init__", init)
+
+    # Create Python properties.
+    for name, prop in props.items():
+        setattr(cls, name, _create_py_property(name, prop.name, prop.type, setter=False))
+
+    # Add an __eq__ method if one doesn't already exist.
+    _add_eq(cls)
+
+    return cls
+
+
+def getter(_fn=None, *, name: Optional[str] = None):
+    """
+    Decorator to indicate a function is a Pulumi property getter.
+
+    name is the Pulumi property name. If not set, the name of the function is used.
+    """
+    def decorator(fn: Callable) -> Callable:
+        # If name isn't specified, use the name of the function.
+        pulumi_name = name if name is not None else fn.__name__
+        if _utils.is_empty_function(fn):
+            @functools.wraps(fn)
+            def get_fn(self):
+                return get(self, pulumi_name)
+            fn = get_fn
+        setattr(fn, _PULUMI_GETTER, True)
+        setattr(fn, _PULUMI_NAME, pulumi_name)
+        return fn
+
+    # See if we're being called as @getter or @getter().
+    if _fn is None:
+        # We're called with parens.
+        return decorator
+
+    # We're called as @getter without parens.
+    return decorator(_fn)
+
+
+def get(self, name: str) -> Any:
+    """
+    Used to get values in Pulumi property getters.
+
+    name is the Pulumi property name.
+    """
+
+    if not name:
+        raise TypeError("Missing name argument")
+    if not isinstance(name, str):
+        raise TypeError("Expected name to be a string")
+
+    values: Optional[Dict[str, Any]] = getattr(self, _VALUES, None)
+
+    if hasattr(type(self), _PULUMI_INPUT_TYPE):
+        return values.get(name) if values is not None else None
+
+    if hasattr(type(self), _PULUMI_OUTPUT_TYPE):
+        cls = type(self)
+
+        # If the class has a _translate_property() method, use it to translate
+        # property names, otherwise, use an identity function.
+        translate = getattr(cls, _TRANSLATE_PROPERTY, None)
+        if not callable(translate):
+            translate = lambda self, prop: prop
```

Comment:
Does this mean that we're translating from Python -> Pulumi names at runtime inside a property getter?

Reply:
It's at runtime, but this is actually translating the other way from Pulumi -> "Python (mapping table)" names, and it's only done for output types.

This is how output types work in this PR:

When unmarshaling the outputs, dictionary keys are translated using the resource's `translate_output_property()` function (as is the case today). If the `dict` is intended to be an output class, we instantiate the class, passing the _translated_ `dict` to the output classes `__init__()`.

The output classes that we generate for our providers inherit from `dict`, so when we pass the _translated_ `dict` to the output class's `__init__()`, it's actually just copying the _translated_ `dict` into itself (if the output class didn't inherit from `dict`, it saves a reference to the _translated_ `dict` as `self._values`). At this point the object has the _translated_ keys, which is what we want to maintain backwards compatibility. Existing programs accessing the values as a `dict` can continue to do so, using the translated key names.

But these objects now have a nicer way of accessing the values, using the new strongly named/typed Python properties that we've added. These properties need to access the values, and since the keys are the translated keys, our provide codegen also generates a `_translate_property()` method on the output class, which does the same things as the resource's `translate_output_property()`: translates Pulumi -> "Python (mapping table)" names. If we see we have a `_translate_property()`, we call it to translate the key to lookup the value.